### ISSUE TITLE
Can we have all ConfigData under $Node.PropertyName?

### ISSUE THREAD

**Issue by mrhockeymonkey** [gaelcolas/Datum#1]
Hey Gael, 

Just some feedback following your presentation since we didn't get to chat properly. I might have missed your intentions but for me as a DSC user I would have thought it simpler and more intuitive to take all the data as you do, apply the ResolutionPrecedence as you do, and then present the *resultant* values as part of $AllNodes. 

A few points why i think this:

* It will let people access variables using $Node.VariableName which is simple and familliar
* It pushes the configuration author to respect the fact that a particular value is "preferred" based on the ResolutionPrecedence (but you may want to allow them to override this as you said)
* It abstracts from the structure of the data which is quite complex, what i mean by this is a lot of data in different formats from different sources is collected. This is likely to result in a complex structure, you yourself had to index into a fair few variables to access the value you wanted. If you flattened this structure respecting the ResolutionPrecedence then it would ease the burden of the author having to figure out where his values are
* It will be less attractive for authors to make exceptions at a whim
* Finally its more fitting with the definition of Datum. I would say that each $Node is the "datum". I wouldn't term a complex structure as "datum", that is more "data". 

Just some thoughts, otherwise i really like the idea. Don't think ill get much of a chance to use this inside my black box but maybe this will come in handy when i start working on some personal DSC projects.

**Comment by gaelcolas** [comment]
Hey @mrhockeymonkey 

This is an interesting input, although it's not _directly_ possible... Let me explain why first.

The benefit of having different layers allows to have an 'override-only' abstraction layer structure, but there's also cases where you want to not only override, but also merge data (that means using the lower layer and add up what each layer defines).
For instance you want to add packages based on their Services, but also, Location, environment and Node specific. Instead of replacing, you want to inherit and merge unique values, so the behaviour you define would not work, as that merging behaviour is defined **at the property resolution stage**.
`Resolve-Datum -SearchBehavior 'AllValues' -...`

When you say that I had to index into a fair few number of property, that was only to illustrate the resulting data structure, and a user would not need it most cases (At least, in my view, not for DSC Configuration Data usage).
They would only resolve the property name, like this:
`Resolve-Datum 'RoleName\ConfigProperty'`

And should rarely need to structure the information too much under the Layer (i.e. Role here), because, well, this is why you have layers. And the Node override definition, for instance, by having no abstraction layer, should probably define any Node specific config at root level...
It'll be more clear when  I do a full fledged example, but DSCConfigurationData works the same way.

So in short, what I meant is that the flexibility of the tool, puts the responsibility on the user(s), to follow a structure that works for them.

That said, there's two ideas that your comments lead to, that should be explored.
It's possible to 'resolve' all layers that a Node is affected by (or depend on), and put them under $Node so that you have:
```PowerShell
$Node.Service
$Node.Site
$Node.Environment
```
Then still resolve the property through the desired node behaviour. That sounds, possible, and maybe not too difficult.

The other solution would be to define the merge-mode behaviour, in an external data, so that it could be enforced from a meta definition...
will sleep on it...

**Comment by gaelcolas** [comment]
Actually, something else triggered some thinking that join this somehow.

In my current demo, I _mount_ the data in `$ConfigurationData.Datum` and access it via:
```
Resolve-Datum -searchPaths $yml.ResolutionPrecedence `
              -DatumStructure $datum `
              -PropertyPath 'ExampleProperty1' `
              -SearchBehavior 'AllValues'
```

But I could abstract that path resolution via ScriptProperty within the `$Node` object, effectively simulating the data being _mounted_ under $Node.
`$Node.PropertyName` -> Calls `Resolve-Datum -DatumStructure $datum -PropertyPath 'PropertyName'`

This does not cater for the case where a merge behaviour is required, but that could be done with:
 - in the configuration script (like `$Node.PropertyName | Select -Unique`)
 - the other way
 - a ScriptMethod

This is very interesting in the Context of Re-usability of Script [Configuration](https://github.com/powershell/dscconfigurations) where I extracted the idea from...

Most of the example configs uses in their example a lookup to `$Node.Property`, so that the Configuration Data can be relatively simple, with no tooling, within a single PSD1 file.
Simulating this behaviour could join towards a standard that means the default config can be re-used, while still allowing config data overriding via Datum: i.e. `$Node.ScriptConfigName.Property`

**Comment by gaelcolas** [comment]
The principle of this is shown in Demo3 (it's actually much harder than I thought).
I need to wire this with the actual function and think a bit more thoroughly of an implementation, but the brick is laid out.

**Comment by gaelcolas** [comment]
I know it's an old one, but yeah now you can, when you do an 'RSOP' (For resultant set of Policy), using `$Node = Get-DatumRSOP $Node`. (It sort of add all missing property to your core `$Node` object)
It's not well documented, but you can see it in action in the gaelcolas/DscInfraSample repo. The RSOP are converted as Yaml files so that you can Diff between commits.

You can then change the RootConfiguration.ps1 to

**Comment by grinnovationlab-prod** [comment]
Sorry haven't had time to write a proper reply. I think when I opened this issue I misunderstood the aim of this module. I've been kicking the tires a bit more today and I'm a bit more on the wavelength now. I'm going to work this into a demo and see how it fares for us (you comment was VERY well timed that I happen to be looking at dsc again recently). I'm happy for you to close this one off. Thanks for the info.

**Comment by mrhockeymonkey** [comment]
Sorry haven't had time to write a proper reply. I think when I opened this issue I misunderstood the aim of this module. I've been kicking the tires a bit more today and I'm a bit more on the wavelength now. I'm going to work this into a demo and see how it fares for us (you comment was VERY well timed that I happen to be looking at DSC again recently). 

Thanks for all the work on this. (closing off)

**Comment by gaelcolas** [comment]
Sounds good @mrhockeymonkey :)
Missed you at PSConfEU!

Do have a look at gaelcolas/DscInfraSample, this is where an example of pipeline is shown. You can also run the AutomatedLab/DscWorkshop that helps you create a full blown environment with a VSTS build pipeline compiling MOFs based on Datum...

If you need more help, feel free to post an issue in any of those project... Also, I know a consultant...  :-D
